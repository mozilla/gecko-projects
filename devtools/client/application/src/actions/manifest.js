/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { services } = require("../modules/services");
const {
  FETCH_MANIFEST_FAILURE,
  FETCH_MANIFEST_START,
  FETCH_MANIFEST_SUCCESS,
  RESET_MANIFEST,
} = require("../constants");

function fetchManifest() {
  return async (dispatch, getState) => {
    dispatch({ type: FETCH_MANIFEST_START });
    const { manifest, errorMessage } = await services.fetchManifest();

    if (!errorMessage) {
      dispatch({ type: FETCH_MANIFEST_SUCCESS, manifest });
    } else {
      dispatch({ type: FETCH_MANIFEST_FAILURE, error: errorMessage });
    }
  };
}

function resetManifest() {
  return { type: RESET_MANIFEST };
}

module.exports = {
  fetchManifest,
  resetManifest,
};
