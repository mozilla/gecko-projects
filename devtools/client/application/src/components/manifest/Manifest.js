/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const {
  article,
  h1,
  table,
  tbody,
} = require("devtools/client/shared/vendor/react-dom-factories");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);
const { l10n } = require("../../modules/l10n");

const ManifestItem = createFactory(require("./ManifestItem"));
const ManifestIssueList = createFactory(require("./ManifestIssueList"));
const ManifestSection = createFactory(require("./ManifestSection"));

/**
 * A canonical manifest, splitted in different sections
 */
class Manifest extends PureComponent {
  static get propTypes() {
    // TODO: Use well-defined types
    //       See https://bugzilla.mozilla.org/show_bug.cgi?id=1576881
    return {
      icons: PropTypes.array.isRequired,
      identity: PropTypes.array.isRequired,
      presentation: PropTypes.array.isRequired,
      validation: PropTypes.array.isRequired,
    };
  }

  renderIssueSection() {
    const { validation } = this.props;
    const shouldRender = validation && validation.length > 0;

    return shouldRender
      ? ManifestSection(
          {
            key: `manifest-section-0`,
            title: l10n.getString("manifest-item-warnings"),
          },
          ManifestIssueList({ issues: validation })
        )
      : null;
  }

  renderItemSections() {
    const { identity, icons, presentation } = this.props;

    const manifestMembers = [
      { localizationId: "manifest-item-identity", items: identity },
      { localizationId: "manifest-item-presentation", items: presentation },
      { localizationId: "manifest-item-icons", items: icons },
    ];

    return manifestMembers.map(({ localizationId, items }, index) => {
      return ManifestSection(
        {
          key: `manifest-section-${index + 1}`,
          title: l10n.getString(localizationId),
        },
        // NOTE: this table should probably be its own component, to keep
        //       the same level of abstraction as with the validation issues
        // Bug https://bugzilla.mozilla.org/show_bug.cgi?id=1577138
        table(
          {},
          tbody(
            {},
            // TODO: handle different data types for values (colors, imagesâ€¦)
            //       See https://bugzilla.mozilla.org/show_bug.cgi?id=1575529
            items.map(item => {
              const { key, value } = item;
              return ManifestItem({ label: key, key: key }, value);
            })
          )
        )
      );
    });
  }

  render() {
    return article(
      {},
      Localized(
        {
          id: "manifest-view-header",
        },
        h1({ className: "app-page__title" })
      ),
      this.renderIssueSection(),
      this.renderItemSections()
    );
  }
}

// Exports
module.exports = Manifest;
