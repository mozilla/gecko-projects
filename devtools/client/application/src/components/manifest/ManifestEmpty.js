/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const {
  article,
  h1,
  p,
} = require("devtools/client/shared/vendor/react-dom-factories");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);

/**
 * This component displays help information when no manifest is found for the
 * current target.
 */
class ManifestEmpty extends PureComponent {
  render() {
    return article(
      { className: "manifest-empty" },
      Localized(
        {
          id: "manifest-view-header",
        },
        h1({ className: "app-page__title" })
      ),
      Localized({ id: "manifest-non-existing" }, p({}))
    );
  }
}

// Exports
module.exports = ManifestEmpty;
