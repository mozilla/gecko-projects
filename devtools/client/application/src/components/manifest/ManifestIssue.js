/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const {
  img,
  li,
  span,
} = require("devtools/client/shared/vendor/react-dom-factories");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);

const { MANIFEST_ISSUE_LEVELS } = require("../../constants");

/**
 * A Manifest validation issue (warning, error)
 */
class ManifestIssue extends PureComponent {
  static get propTypes() {
    return {
      className: PropTypes.string,
      level: PropTypes.oneOf(Object.values(MANIFEST_ISSUE_LEVELS)).isRequired,
      message: PropTypes.string.isRequired,
      // NOTE: we are currently ignoring the 'type' field that platform adds to
      //       errors
    };
  }

  getIconData(level) {
    switch (level) {
      case MANIFEST_ISSUE_LEVELS.WARNING:
        return {
          src: "chrome://devtools/skin/images/alert-small.svg",
          localizationId: "icon-warning",
        };
      case MANIFEST_ISSUE_LEVELS.ERROR:
      default:
        return {
          src: "chrome://devtools/skin/images/error-small.svg",
          localizationId: "icon-error",
        };
    }
  }

  render() {
    const { level, message, className } = this.props;
    const icon = this.getIconData(level);

    return li(
      { className: `manifest-warning ${className ? className : ""}` },
      Localized(
        { id: icon.localizationId, attrs: { alt: true, title: true } },
        img({
          className: `manifest-issue__icon manifest-issue__icon--${level}`,
          src: icon.src,
        })
      ),
      span({}, message)
    );
  }
}

// Exports
module.exports = ManifestIssue;
