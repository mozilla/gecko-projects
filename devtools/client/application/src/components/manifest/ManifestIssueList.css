.manifest-issues {
  list-style-type: none;
  padding-inline-start: 0;
  display: grid;
  grid-template-columns: auto 1fr;
  grid-gap: var(--base-unit);
}

.manifest-issues__item {
  display: contents;
}
