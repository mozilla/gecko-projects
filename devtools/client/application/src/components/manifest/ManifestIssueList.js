/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const { ul } = require("devtools/client/shared/vendor/react-dom-factories");

const { MANIFEST_ISSUE_LEVELS } = require("../../constants");
const ManifestIssue = createFactory(require("./ManifestIssue"));

/**
 * A collection of manifest issues (errors, warnings)
 */
class ManifestIssueList extends PureComponent {
  static get propTypes() {
    // TODO: Use well-defined types
    //       See https://bugzilla.mozilla.org/show_bug.cgi?id=1576881
    return {
      issues: PropTypes.array.isRequired,
    };
  }

  // group the errors by level, and order by severity
  groupIssuesByLevel() {
    const { issues } = this.props;

    const errors = issues.filter(x => x.level === MANIFEST_ISSUE_LEVELS.ERROR);
    const warnings = issues.filter(
      x => x.level === MANIFEST_ISSUE_LEVELS.WARNING
    );

    return [errors, warnings];
  }

  render() {
    const groups = this.groupIssuesByLevel().filter(list => list.length > 0);

    return groups.map((list, listIndex) => {
      return ul(
        {
          className: "manifest-issues js-manifest-issues",
          key: `issuelist-${listIndex}`,
        },
        list.map((issue, issueIndex) =>
          ManifestIssue({
            className: "manifest-issues__item",
            key: `issue-${issueIndex}`,
            ...issue,
          })
        )
      );
    });
  }
}

// Exports
module.exports = ManifestIssueList;
