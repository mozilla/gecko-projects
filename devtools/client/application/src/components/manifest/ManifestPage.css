/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.manifest-page {
	height: 100vh;
  padding: 0 2rem;
  display: grid;
  -moz-user-select: none;

  /* slipt up in components in https://bugzilla.mozilla.org/show_bug.cgi?id=1566011 */
  align-items: center;
  justify-content: center;
  font-size: var(--title-10-font-size);
  color: var(--theme-toolbar-color);
}
