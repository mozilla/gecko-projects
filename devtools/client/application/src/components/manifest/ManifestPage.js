/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const {
  section,
} = require("devtools/client/shared/vendor/react-dom-factories");

const ManifestLoader = createFactory(require("../manifest/ManifestLoader"));

const Manifest = createFactory(require("./Manifest"));
const ManifestEmpty = createFactory(require("./ManifestEmpty"));

class ManifestPage extends PureComponent {
  render() {
    // TODO: needs to be replaced with data from Redux
    const data = {
      validation: [
        { level: "warning", message: "Icons item at index 0 is invalid." },
        { level: "error", message: "Random JSON error" },
        {
          level: "warning",
          message:
            "Icons item at index 2 is invalid. Icons item at index 2 is invalid. Icons item at index 2 is invalid. Icons item at index 2 is invalid.",
        },
      ],
      icons: [],
      identity: [
        {
          key: "name",
          value:
            "Name is a verrry long name and the name is longer tha you thinnk because it is loooooooooooooooooooooooooooooooooooooooooooooooong",
        },
        {
          key: "short_name",
          value: "Na",
        },
      ],
      presentation: [
        { key: "display", value: "browser" },
        { key: "orientation", value: "landscape" },
        { key: "start_url", value: "root" },
        { key: "theme_color", value: "#345" },
        { key: "background_color", value: "#F9D" },
      ],
    };

    const isManifestEmpty = !data;

    return section(
      {
        className: `app-page ${isManifestEmpty ? "app-page--empty" : ""}`,
      },
      ManifestLoader({}),
      isManifestEmpty
        ? ManifestEmpty({})
        : Manifest({
            icons: data.icons,
            identity: data.identity,
            presentation: data.presentation,
            validation: data.validation,
          })
    );
  }
}

// Exports
module.exports = ManifestPage;
