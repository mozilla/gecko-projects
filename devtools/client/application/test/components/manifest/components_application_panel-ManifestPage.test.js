/* Any copyright is dedicated to the Public Domain.
   http://creativecommons.org/publicdomain/zero/1.0/ */

"use strict";

// Import libs
const { shallow } = require("enzyme");
const { createFactory } = require("react");

const ManifestPage = createFactory(
  require("devtools/client/application/src/components/manifest/ManifestPage")
);

/**
 * Test for ManifestPage.js component
 */

describe("ManifestPage", () => {
  it("renders the expected snapshot", () => {
    const wrapper = shallow(ManifestPage({}));
    expect(wrapper).toMatchSnapshot();
  });
});
