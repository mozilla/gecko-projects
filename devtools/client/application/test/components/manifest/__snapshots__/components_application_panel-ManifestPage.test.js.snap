// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManifestPage renders the expected snapshot 1`] = `
<section
  className="app-page "
>
  <Connect(ManifestLoader) />
  <Manifest
    icons={Array []}
    identity={
      Array [
        Object {
          "key": "name",
          "value": "Name is a verrry long name and the name is longer tha you thinnk because it is loooooooooooooooooooooooooooooooooooooooooooooooong",
        },
        Object {
          "key": "short_name",
          "value": "Na",
        },
      ]
    }
    presentation={
      Array [
        Object {
          "key": "display",
          "value": "browser",
        },
        Object {
          "key": "orientation",
          "value": "landscape",
        },
        Object {
          "key": "start_url",
          "value": "root",
        },
        Object {
          "key": "theme_color",
          "value": "#345",
        },
        Object {
          "key": "background_color",
          "value": "#F9D",
        },
      ]
    }
    validation={
      Array [
        Object {
          "level": "warning",
          "message": "Icons item at index 0 is invalid.",
        },
        Object {
          "level": "error",
          "message": "Random JSON error",
        },
        Object {
          "level": "warning",
          "message": "Icons item at index 2 is invalid. Icons item at index 2 is invalid. Icons item at index 2 is invalid. Icons item at index 2 is invalid.",
        },
      ]
    }
  />
</section>
`;
